pinc/wordcheck_engine.inc

```diff
@@ -785,19 +785,24 @@ function load_word_list($path)
         return [];
     }
 
-    // Note: we're unable to use the file() function's FILE_IGNORE_NEW_LINES
-    // parameter since it wasn't added until PHP 5.0.0 *and* because
-    // it only removes \n, not \r\n, so we'd still need to rtrim later
-    $words = file($path);
-    if ($words === false) {
+    // It's faster to load the entire file into memory, remove any trailing
+    // spaces, and then explode it into an array than use file() to load
+    // it into an array and then rtrim() / rtrim_walk() the entries. Word lists
+    // should certainly be small enough that we can load the contents into
+    // memory for a brief period.
+
+    // Note that this function needs to work on files that contain \r\n line
+    // endings since that's what we write out in save_word_list() -- which
+    // precludes us from using file() with the FILE_IGNORE_NEW_LINES flag.
+    $file_contents = file_get_contents($path);
+    if ($file_contents === false) {
         return sprintf(_("Error reading file %s"), $path);
     }
 
-    // trim word list to remove trailing spaces
-    // needed in case site word list has trailing spaces
-    // And if/since the file has \r\n line-ends,
-    // this will also remove the trailing \r.
-    array_walk($words, 'rtrim_walk');
+    // remove any trailing spaces, although there shouldn't be any for
+    // lists saved with save_word_list().
+    $file_contents = preg_replace("/\s+\r*\n/", "\n", $file_contents);
```

Comment:
Isn't this equivalent to "/\s+/" since \s matches space or \r or \n ?

Reply:
Good question. In this case I only want to trim **trailing** whitespaces, so I want to anchor the pattern with a newline. If I just removed `/\s+/` it would remove any duplicate whitespace which might also include whitespace elsewhere. There shouldn't be whitespace elsewhere, but I want to replicate what was happening before.

Reply:
Although your comment just now made me realize that I ought to use `/\s*\r*\n/` so it's clearer that we're correctly handling the case of `\r`s. Right now those are working because they're being consumed by the `\s` which might be confusing later. I'll get that updated now.